### PR TITLE
Migrate to Launch Templates

### PR DESCRIPTION
**Pull request by errordeveloper**
### Description

Towards #198, #320 and #643.

### Checklist

- [x] Code compiles correctly (i.e `make build`)
- [x] Added tests that cover your change (if possible)
- [x] All unit tests passing (i.e. `make test`)
- [x] All integration tests passing (i.e. `make integration-test`)

### CODE REVIEW

File: pkg/cfn/builder/nodegroup.go [hunk 1]
```diff
@@ -190,9 +199,9 @@ func (n *NodeGroupResourceSet) addResourcesForNodeGroup() error {
 		)
 	}
 	ngProps := map[string]interface{}{
-		"LaunchConfigurationName": refLC,
-		"VPCZoneIdentifier":       vpcZoneIdentifier,
-		"Tags":                    tags,
+		"LaunchTemplateName": launchTemplateName,
+		"VPCZoneIdentifier":  vpcZoneIdentifier,
+		"Tags":               tags, // TODO: make part of LT
```
Comment (errordeveloper) [inline pkg/cfn/builder/nodegroup.go]: @christopherhein should we do that, or keep them on the ASG?

**Comment by errordeveloper** [comment]
Still need to test it properly, looks integration tests are failing on first attempt.

**Comment by errordeveloper** [comment]
Now I'm getting this:
```
2019-04-26T15:26:20+01:00 [✖]  AWS::AutoScaling::AutoScalingGroup/NodeGroup: CREATE_FAILED – "Incompatible launch template: The network interface's device index must be zero. (Service: AmazonAutoScaling; Status Code: 400; Error Code: InvalidQueryParameter; Request ID: 3f038e83-682f-11e9-b8b7-57afd91b5253)"
```

@christopherhein any ideas?

File: pkg/cfn/builder/nodegroup.go [hunk 2]
```diff
@@ -107,33 +107,42 @@ func (n *NodeGroupResourceSet) newResource(name string, resource interface{}) *g
...
+		UserData:       n.userData,
 	}
+
 	if api.IsEnabled(n.spec.SSH.Allow) && api.IsSetAndNonEmptyString(n.spec.SSH.PublicKeyName) {
-		lc.KeyName = gfn.NewString(*n.spec.SSH.PublicKeyName)
+		launchTemplateData.KeyName = gfn.NewString(*n.spec.SSH.PublicKeyName)
 	}
 	if n.spec.PrivateNetworking {
-		lc.AssociatePublicIpAddress = gfn.False()
+		launchTemplateData.NetworkInterfaces = []gfn.AWSEC2LaunchTemplate_NetworkInterface{{
```
Comment (errordeveloper) [inline pkg/cfn/builder/nodegroup.go]: So it looks like we shouldn't actually define it here at all.

If so, where is this this parameter supposed to be set?

There are few things in [the docs](https://docs.aws.amazon.com/autoscaling/ec2/userguide/create-launch-template.html), but it's still not clear to me what I'm supposed to do here really.

Comment (errordeveloper) [reply pkg/cfn/builder/nodegroup.go]: Or do I need to set actual `DeviceIndex` parameter?

Comment (christopherhein) [reply pkg/cfn/builder/nodegroup.go]: Looks like you might need to set Device to eth0 ?

Comment (martina-if) [reply pkg/cfn/builder/nodegroup.go]: I tried with 
```
launchTemplateData.NetworkInterfaces = []gfn.AWSEC2LaunchTemplate_NetworkInterface{{
	AssociatePublicIpAddress: gfn.False(),
	DeviceIndex:              gfn.NewInteger(0),
}}
```

but I still got the same error:
```[ℹ]  AWS::CloudFormation::Stack/eksctl-martina-test-cluster-6-nodegroup-ng-efc71c9c: ROLLBACK_IN_PROGRESS – "The following resource(s) failed to create: [NodeGroup]. . Rollback requested by user."
[✖]  AWS::AutoScaling::AutoScalingGroup/NodeGroup: CREATE_FAILED – "Incompatible launch template: The network interface's device index must be zero. (Service: AmazonAutoScaling; Status Code: 400; Error Code: InvalidQueryParameter; Request ID: e39d1400-6b18-11e9-aa8c-a7959d188056)"```

Comment (errordeveloper) [reply pkg/cfn/builder/nodegroup.go]: @christopherhein I don't think there is `Device`, it's `DeviceIndex`...

@martina-if and you've tried that with `privateNetworking: true`?

I actually managed to get past it now, and seeing this:
```
[✖]  AWS::AutoScaling::AutoScalingGroup/NodeGroup: CREATE_FAILED – "Invalid launch template: When a network interface is provided, the security groups must be a part of it. (Service: AmazonAutoScaling; Status Code: 400; Error Code: InvalidQueryParameter; Request ID: 76da1253-6b1c-11e9-a368-39534edcaa0d)"
```

**Comment by errordeveloper** [comment]
We still need decide about tags and, this clearly needs more testing. I don't think it's a good idea to rush it for 0.1.30 today.

File: pkg/cfn/builder/nodegroup.go (hunk 2 above)
Comment (errordeveloper) [reply pkg/cfn/builder/nodegroup.go]: Looks like we have a working thing now, let's come back to this after 0.1.30 is out.

**Comment by errordeveloper** [comment]
TODOs:
- [ ] decide on tags
- [ ] fix unit tests

**Pull request merged**